src/services/getExpectedApy.ts

```diff
@@ -11,16 +11,48 @@ export const getExpectedApy = (
   vt: number, // variable token balance
   margin: number, // margin until the current timestamp (margin + accrued cashflow)
   predictedApr: number, // the predicted variable APY between current and end
+  rateOracleID: number, // the rate oracle ID that suggests what type of APY is used (linear vs. compounding)
 ): [number, number] => {
-  // (rate[end]/rate[current])^(YEAR / (end-current)) - 1 = APY(current, end)
-  // rate[end]/rate[current] = (APY(current, end) + 1)^((end - current)/YEAR)
-  // rate[end]/rate[current] - 1 = (APY(current, end) + 1)^((end - current)/YEAR) - 1
-  const vf = (1 + predictedApr) ** getAnnualizedTime(current, end) - 1;
+  if (end <= current) {
+    return [0, 0];
+  }
+
+  let vf = 0;
+
+  switch (rateOracleID) {
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7: {
+      // (rate[end]/rate[current])^(YEAR / (end-current)) - 1 = APY(current, end)
+      // rate[end]/rate[current] = (APY(current, end) + 1)^((end - current)/YEAR)
+      // rate[end]/rate[current] - 1 = (APY(current, end) + 1)^((end - current)/YEAR) - 1
+      vf = (1 + predictedApr) ** getAnnualizedTime(current, end) - 1;
+
+      break;
+    }
+
+    case 8: {
+      // (rate[end]/rate[current]-1) * (YEAR / (end - current)) = APY(current, end)
+      // rate[end]/rate[current] = APY(current, end) * (end - current) / YEAR + 1
+      // rate[end]/rate[current] - 1 = APY(current, end) * (end - current) / YEAR
+      vf = predictedApr * getAnnualizedTime(current, end);
+
+      break;
+    }
+
+    default: {
+      throw new Error('Unrecognized protocol');
+    }
+  }
 
   // cashflow = uft * (end - current) / YEAR * 0.01 + vt * (rate[end]/rate[current] - 1)
   const ecs = uft * getAnnualizedTime(current, end) * 0.01 + vt * vf;
```

Comment:
just to double check, this calculation outputs the expected cashflow that the position as a whole would accrue from now to end if the new swap were to be done in this case, if that's the case worth detailing that in the comments i think for readability

Reply:
also, haven't been through the math behind why we can just aggregate unbalanced fixed tokens and use them to calculate the expected cashflow even if these unbalanced fixed tokens were generated at different points in time, is the math 100% correct here?

Reply:
yes, they can be aggregated because we want the APY from now until the end.